testsys: Add declarative testing through `Test.toml`

ecpullen opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**

Closes # N/A

**Description of changes:**

Enables configuring testing through `Test.toml`. There are now 5 levels of configuration.

```
# The test configuration is broken up into different levels depending on the variant.
# Each configuration's value will be selected in the following order:
# Arguments from the command line
# Environment variables from the command line
# '{variant}' configuration in `Test.toml`
# '{family}-{flavor}' configuration in `Test.toml`
# '{family}' configuration in `Test.toml`
# '{platform}-{flavor}' configuration in `Test.toml`
# '{platform}' configuration in `Test.toml`
# TestSys default values
```

**Testing done:**

Testing workflow over many different configurations.
* Create a sample config file
* Run `cargo make test`
* Use `kubectl describe test <TEST-NAME>` and `kubectl describe resource <RESOURCE-NAME>` to see the `configuration`
* Verify that the values in the `configuration` match the values in the config file.

I tested this with a variety of different configurations making sure that each field was selected properly for each possible field. I then tested that the configuration values were placed in the correct spots in the crds.

I also tested `Test.toml.example` with the redacted fields filled in to make sure that the documentation was correct. 


**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

bcressey commented on tools/testsys/Test.toml.example:
```diff
@@ -0,0 +1,78 @@
+# needed for testing
+repo = "default"
+
+# Configuration for all `aws-k8s` variants testing.
+[aws-k8s]
+# A single assume role can be used to test all `aws-k8s` variants in a separate
+# testing account.
+assume_role = "arn:aws:iam::<account_id>:role/<role_name>
```
Does the use of this role follow the same nested approach as `Infra.toml`? Meaning:
* first we find environment creds or a profile
* then we assume the global role (if any)
* then we assume the regional role (if any)
* then we assume the test role here (if any)

(nit: I'd also prefer calling the field `role` to align with `Infra.toml`.)

Also, if we don't need to support different AWS accounts per variant under test, we could potentially add a `test_role` to the global and regional levels in `Infra.toml` instead, just to consolidate the AWS specifics.

ecpullen replied:
The assume role here is the role that is assumed by each testsys agent not by the testsys cli, but the nested approach is implemented in the individual agents.

bcressey commented on tools/testsys/Test.toml.example:
```diff
@@ -0,0 +1,78 @@
+# images need to be copied in a registry.
+e2e_repo_registry = "<account>.dkr.ecr.cn-north-1.amazonaws.com.cn"
+
+# If testing using a kind cluster, aws credentials need to be passed as a k8s secret.
+secrets = {"awsCreds" = "myAwsCredentials"}
```
What is the interaction between this secret and the role credentials we're meant to assume? Does the secret get automatically created and refreshed?

ecpullen replied:
These are secrets that live in the k8s cluster. If the k8s cluster is a kind cluster, the nodes will not have aws credentials. This secret provides a way for those agents to gain aws credentials for using the aws sdk. The secret is applied to the cluster using `testsys add secret map ...`.

arnaldo2792 commented on tools/testsys-config/src/lib.rs:
```diff
@@ -0,0 +1,129 @@
+}
+
+impl TestConfig {
+    /// Deserializes an TestConfig from a given path
```
Nit:

```suggestion
    /// Deserializes a TestConfig from a given path
```

Same for all occurrences

arnaldo2792 commented on tools/testsys/src/aws_resources.rs:
```diff
@@ -286,7 +319,30 @@ impl AwsK8s {
+                                |e2e_registry| {
+                                    base64::encode(format!(
+                                        r#"buildImageRegistry: {}
+dockerGluster: {}
```
Is this intended, or should it be `dockerCluster`?

ecpullen replied:
It is intended.

arnaldo2792 commented on tools/testsys-config/src/lib.rs:
```diff
@@ -93,6 +102,42 @@ pub struct Test {
+#[serde(deny_unknown_fields)]
+#[serde(rename_all = "kebab-case")]
+pub struct AwsK8sConfig {
+    /// The rle that should be assumed by all `aws-k8s` variants
```
```suggestion
    /// The role that should be assumed by all `aws-k8s` variants
```

arnaldo2792 commented on tools/testsys-config/src/lib.rs:
```diff
@@ -93,6 +102,42 @@ pub struct Test {
+    /// for testing Bottlerocket on ipv4 and ipv6 clusters.
+    #[serde(default)]
+    pub cluster_names: Vec<String>,
+    /// The instance type that instance types should be launched with
```
Should this be?

```suggestion
    /// The instance type that instances should be launched with
```

arnaldo2792 commented on tools/testsys/src/run.rs:
```diff
@@ -44,6 +44,10 @@ pub(crate) struct Run {
     ami_input: String,
 
+    /// The repo containing Sonobuoy e2e images
+    #[clap(long, env = "TESTSYS_E2E_REPO_REGISTRY")]
```
Should this be namespaced for k8s/sonobuoy? i. e. `TESTSYS_<HERE>_E2E_REPO_REGISTRY`

arnaldo2792 commented on tools/testsys/src/run.rs:
```diff
@@ -171,16 +169,70 @@ impl Run {
+                    .collect();
+                if target_cluster_names.len() == 0 {
+                    target_cluster_names.push(None);
+                }
```
I feel this is a little hacky, and instead the code should reflect the intent of using defaults when no cluster names were passed. With this it is less obvious that `None` actually means "use default values".

arnaldo2792 commented on tools/testsys/src/aws_resources.rs:
```diff
@@ -286,7 +319,30 @@ impl AwsK8s {
+                                    base64::encode(format!(
+                                        r#"buildImageRegistry: {}
+dockerGluster: {}
+dockerLibraryRegistry: {}
+e2eRegistry: {}
+e2eVolumeRegistry: {}
+gcRegistry: {}
+gcEtcdRegistry: {}
+promoterE2eRegistry: {}
+sigStorageRegistry: {}"#,
+                                        e2e_registry,
+                                        e2e_registry,
+                                        e2e_registry,
+                                        e2e_registry,
+                                        e2e_registry,
+                                        e2e_registry,
+                                        e2e_registry,
+                                        e2e_registry,
+                                        e2e_registry
+                                    ))
+                                },
+                            ),
```
You could do this instead:

```suggestion
dockerLibraryRegistry: {e2e_registry}
e2eRegistry: {e2e_registry}
e2eVolumeRegistry: {e2e_registry}
gcRegistry: {e2e_registry}
gcEtcdRegistry: {e2e_registry}
promoterE2eRegistry: {e2e_registry}
sigStorageRegistry: {e2e_registry}"#,
                                        e2e_registry,
```

arnaldo2792 commented on tools/testsys-config/src/lib.rs:
```diff
@@ -106,7 +108,7 @@ pub struct Test {
 #[serde(rename_all = "kebab-case")]
 pub struct AwsK8sConfig {
-    /// The rle that should be assumed by all `aws-k8s` variants
+    /// The role that should be assumed by all `aws-k8s` variants
```
This shouldn't be in this commit, since the commit is related to `aws-ecs`.

arnaldo2792 commented on tools/testsys-config/src/lib.rs:
```diff
@@ -138,6 +140,36 @@ pub struct AwsK8sVariantConfig {
+#[serde(deny_unknown_fields)]
+#[serde(rename_all = "kebab-case")]
+pub struct AwsEcsConfig {
+    /// The role tat should be assumed by all `aws-k8s` variants
```
Nit

```suggestion
    /// The role that should be assumed by all `aws-ecs` variants
```

arnaldo2792 commented on tools/testsys-config/src/lib.rs:
```diff
@@ -138,6 +140,36 @@ pub struct AwsK8sVariantConfig {
+pub struct AwsEcsConfig {
+    /// The role tat should be assumed by all `aws-k8s` variants
+    pub assume_role: Option<String>,
+    /// A set of configurations for specific `aws-k8s` variants
```
`s/aws-k8s/aws-ecs` in this commit?

arnaldo2792 commented on tools/testsys/src/run.rs:
```diff
@@ -146,7 +146,7 @@ impl Run {
                 &self
                     .repo
-                    .unwrap_or(test.repo.unwrap_or("default".to_string())),
+                    .unwrap_or_else(|| test.repo.unwrap_or_else(|| "default".to_string())),
```
If this are cleaning changes, I'd prefer them to be in the commit that introduce the change (i. e. `target_cluster_names.len() == 0` was introduced in a different commit)

arnaldo2792 commented on tools/testsys/src/run.rs:
```diff
@@ -250,15 +250,42 @@ impl Run {
+                    aws_config.variant.get(&format!(
+                        "{}-{}",
+                        self.arch,
+                        self.variant.replace('.', "")
```
A note on why `.` is removed from i. e. `aws-k8s-1.21`?

arnaldo2792 commented on tools/testsys/Test.toml.example:
```diff
@@ -0,0 +1,78 @@
+# testing account.
+assume_role = "arn:aws:iam::<account_id>:role/<role_name>
+
+# To test sonobuoy certified conformance in other AWS partitions, sonobuoy e2e
+# images need to be copied in a registry.
+e2e_repo_registry = "<account>.dkr.ecr.cn-north-1.amazonaws.com.cn"
```
Maybe you can add a comment for an example of other regions?

arnaldo2792 commented on tools/testsys/Test.toml.example:
```diff
@@ -0,0 +1,78 @@
+# images need to be copied in a registry.
+e2e_repo_registry = "<account>.dkr.ecr.cn-north-1.amazonaws.com.cn"
+
+# If testing using a kind cluster, aws credentials need to be passed as a k8s secret.
```
`s/aws/AWS`

Also an example of the format for the AWS credentials?

arnaldo2792 commented on tools/testsys/Test.toml.example:
```diff
@@ -0,0 +1,78 @@
+# removed from the variant.
+[aws-k8s.variant.x86_64-aws-k8s-121]
+
+# Testing may need to be tested on different clusters to test certain behavior. A test will be run
+# on each cluster specified.
```
I think this should be sufficient? 

```suggestion
# A test will run on each cluster specified in this list
```

arnaldo2792 commented on tools/testsys/Test.toml.example:
```diff
@@ -0,0 +1,78 @@
+secrets = {"awsCreds" = "myAwsCredentials"}
+
+# A custom kube conformance image can be used for sonobuoy conformance testing.
+kube_conformance_image = "kube_conformance_image_uri,
```
Nit:

```suggestion
kube_conformance_image = "kube_conformance_image_uri"
```

arnaldo2792 commented on tools/testsys/Test.toml.example:
```diff
@@ -0,0 +1,78 @@
+assume_role = "arn:aws:iam::<account_id>:role/<role_name>
+
+# To test sonobuoy certified conformance in other AWS partitions, sonobuoy e2e
+# images need to be copied in a registry.
```
```suggestion
images need to be available in a registry in the targeted region
```

Same for other occurrences.

arnaldo2792 commented on tools/testsys/Test.toml.example:
```diff
@@ -0,0 +1,78 @@
+# This is an example testing configuration for testsys, the tool that
```
For the lack of a better place for this comment, don't we need documentation about this file and how it will be used by `cargo make test`?

bcressey commented on tools/testsys/Test.toml.example:
```diff
@@ -0,0 +1,78 @@
+
+# To test sonobuoy certified conformance in other AWS partitions, sonobuoy e2e
+# images need to be copied in a registry.
+e2e_repo_registry = "<account>.dkr.ecr.cn-north-1.amazonaws.com.cn"
```
This feels problematic in that we might need to expand this along a regional matrix, and this structure isn't really set up to support that, unlike `Infra.toml`.

It might be better in `Infra.toml` instead.

ecpullen replied:
I opted to wait on this change until after I reworked the rest of the pr.

bcressey commented on tools/testsys/Test.toml.example:
```diff
@@ -0,0 +1,78 @@
+secrets = {"awsCreds" = "myAwsCredentials"}
+
+# Variant based configuration can be set using [aws-ecs.variant.<arch>-<variant>].
+[aws-ecs.variant.x86_64-aws-ecs-1]
```
This is unsatisfying because we've already named part of the variant (the family) so effectively this parses as:
```
{{ family }}.variant.{{ arch }}{{ variant }}
```

Which is just kind of a jumble.

I wonder if it'd be better to support composition and some sort of defined precedence order, e.g.
```
# platform, base level
# we're using hardware-like identifiers which are platform-specific
[aws.x86_64]
instance_type = ..."

# ... can be overridden by "family" specifier
[aws-k8s.x86_64]
instance_type = ..."

# ... can be overridden by the fully qualified variant name
[aws-k8s-1.21-nvidia.x86_64]
instance_type = ..."

# might be nice to have, some sort of "platform plus flavor" specifier
[aws-nvidia.x86_64]
instance_type = "..."

[aws-nvidia.aarch64]
instance_type = "..."
```

To wind that back, I'd say the minimum is:
* a platform level override, e.g. to set a default instance type for any `aws` variant
* a fully-qualified variant override, to override the instance type on a per-variant basis

bcressey commented on tools/testsys/Test.toml.example:
```diff
@@ -0,0 +1,78 @@
+
+# Testing may need to be tested on different clusters to test certain behavior. A test will be run
+# on each cluster specified.
+cluster_names = ["myCluster", "anotherTestCluster"]
```
A challenge here is that there's not really a place for a platform- or family-level default, because we're naming specific clusters which are unlikely to work for all variants.

I'd recommend leaning into an opinionated, convention-over-configuration approach and supporting templates so that only exceptions need to be documented:
```suggestion
[aws]
cluster_names = ["{{ variant }}-{{ arch }}"]

[aws-k8s]
cluster_names = ["{{ variant }}-{{ arch }}", "{{ variant }}-{{ arch }}-ipv6"]
```

bcressey commented on tools/testsys/Test.toml.example:
```diff
@@ -0,0 +1,78 @@
+# The default instance type to be used is m5.large for x86_64 and m6g.large for aarch64, but can be 
+# set here. This is useful while testing NVIDIA variants, since they require instance types with 
+# support for NVIDIA GPUs.
+instance_type = "p3.2xlarge"
```
With the suggested composition model, we wouldn't need hard-coded defaults.

ecpullen commented:
^ Contains a large refactor of changes. `Test.toml` schema has been significantly changed to allow a more concise config.

arnaldo2792 commented on TESTING.md:
```diff
@@ -118,6 +118,11 @@ The Bottlerocket components are found in the `testsys-bottlerocket-aws` Kubernet
 
+### Configuration
+
+There are many arguments that can be configured via environment variables with `cargo make`; however, it is possible to create a configuration file instead.
+Check out the [example config file](tools/testsys/Test.toml.example) for a sample `Test.toml` file.
```
In other places where a configuration file is used, there is documentation that explains a few fields of the configuration file, and what happens when they are set, `Test.toml` should have something similar, and you should provide snippets explaining what will happen when used.

ecpullen replied:
Do we need more documentation than the example config has?

arnaldo2792 replied:
I won't argue about this since it seems like we already link to example configuration files to explain the different sections they have. We could consider not doing so, and do what other open source projects do: provide a document to explain the purpose of each section.

arnaldo2792 reviewed (commented):
There is some clean up going on, I'll post this comment so that it isn't deleted with the force push.

ecpullen committed:
testsys: Update to v0.0.2

This update was partially completed in #2392. Image locations have been
updated and e2e repo config for sonobuoy agents has been added.

ecpullen committed:
testsys: Introduce `Test.toml` for test configs

Add declarative testing for `cargo make test` this allows users to setup
configuration once and use `cargo make -e BUILDSYS_ARCH=<MY-ARCH> -e
BUILDSYS_VARIANT=<MY-VARIANT> test` without any other needed variables.

ecpullen commented:
^ Fixes some rebasing issues and code cleanup.

arnaldo2792 commented on tools/testsys-config/src/lib.rs:
```diff
@@ -0,0 +1,399 @@
+        keys.push(format!("{}-{}", variant.platform(), flavor));
+    }
+    keys.push(variant.platform().to_string());
+    keys
```
```suggestion
    let mut family_flavor: String = "".to_string();
    let mut platform_flavor: String = "".to_string();

    if let Some(flavor) = variant.variant_flavor {
        family_flavor = format!("{}-{}", variant.family(), flavor);
        platform_flavor = format!("{}-{}", variant.platform(), flavor);
    }

    let mut keys = vec![variant.to_string(),
      family_flavor,
      variant.family().to_string(),
      platform_flavor,
      variant.platform().to_string()
    ];

    keys.retain(|s| !s.is_empty());
    keys
```

arnaldo2792 commented on tools/testsys-config/src/lib.rs:
```diff
@@ -0,0 +1,399 @@
+            kube_conformance_image: self.kube_conformance_image.or(other.kube_conformance_image),
+            e2e_repo_registry: self.e2e_repo_registry.or(other.e2e_repo_registry),
+        }
+    }
```
I've seen in other places in the repo that it is preferred to assign values to variables, and then create the struct at the end of the function, instead of having inline ifs to choose the value to be set.

arnaldo2792 commented on tools/testsys/Test.toml.example:
```diff
@@ -0,0 +1,78 @@
+eks-resource-agent-image = "public.ecr.aws/bottlerocket-test-system/eks_resource_agent:v0.0.2"
+
+# The test configuration is broken up into different levels depending on the variant.
+# Each configuration's value will be selected in the following order:
```
Probably "selected" isn't the right term here, probably you should word this like "the configurations have the following precedence". I still think it is worth having an example snippet of configuration in the markdown file that explains what values will be used from each section to help users understand the precedence when two settings are provided at different levels.

arnaldo2792 commented on tools/testsys/Test.toml.example:
```diff
@@ -0,0 +1,78 @@
+# The registry containing alternate TestSys agent images
+testsys-images-registry = "public.ecr.aws/bottlerocket-test-system"
+
+# The uri for the EKS resource agent that should be used.
```
Nit: `s/uri/URI`

arnaldo2792 commented on tools/testsys/Test.toml.example:
```diff
@@ -0,0 +1,78 @@
+# cluster-names: 
+#     For `aws-k8s` variants testing will be completed over all clusters listed.
+#     For `aws-ecs` variants, only the first cluster name will be used.
+#     Cluster naming supports handlebars for `arch` and `variant` (`{{arch}}-{{variant}}`).
```
```suggestion
#     Cluster naming supports templated strings, and both `arch` and `variant` are provided as variables (`{{arch}}-{{variant}}`).
```

arnaldo2792 commented on tools/testsys-config/src/lib.rs:
```diff
@@ -0,0 +1,399 @@
+}
+
+impl GenericConfig {
+    pub fn config<S1>(&self, arch: S1) -> GenericVariantConfig
```
I would use a different name, elsewhere in the PR I was confused since this method, attribute in the struct and the variable have the same name `vec![config.config(&arch), config.config]`. Probably `for_arch()` will be more idiomatic?

arnaldo2792 commented on tools/testsys-config/src/lib.rs:
```diff
@@ -0,0 +1,399 @@
+            .join("Test.lock"))
+    }
+
+    pub fn reduced_config<S1>(
```
Nit: add comments to at least public methods. There are comments in other sources of the repo that explain what each method does so it will be nice to keep the same bar.

arnaldo2792 commented on tools/testsys/src/aws_resources.rs:
```diff
@@ -81,15 +106,23 @@ impl AwsK8s {
             .to_string();
-        let eks = self.eks_crd(testsys_images)?;
-        let ec2 = self.ec2_crd(testsys_images, Some(ami))?;
+        let eks = self.eks_crd(cluster_name.clone(), testsys_images)?;
```
Do you need to clones `cluster_name` twice? It seems like in the method where `cluster_name` is used, the string is cloned again. Wouldn't it be sufficient to pass a reference here and clone down the road?

arnaldo2792 commented on tools/testsys/src/aws_resources.rs:
```diff
@@ -660,6 +730,22 @@ impl Migration for AwsEcs {
     }
 }
 
+// struct MigrationsConfig {
+//     /// The TUF repos for migration testing. If no TUF repos are used, the default Bottlerocket
+//     /// repos will be used
+//     pub(crate) tuf_repo: Option<TufRepoConfig>,
+//     /// The starting version for migration testing
+//     pub(crate) starting_version: Option<String>,
+//     /// The AMI id of the starting version for migration testing
+//     pub(crate) starting_image_id: Option<String>,
+//     /// The short commit SHA of the starting version
+//     pub(crate) migrate_starting_commit: Option<String>,
+//     /// The target version for Bottlerocket migrations
+//     pub(crate) migrate_to_version: Option<String>,
+//     /// Additional capabilities that need to be enabled on the agent's pods
+//     pub(crate) capabilities: Option<Vec<String>>,
+// }
+
```
Is this required? or can it be removed?

ecpullen replied:
That was work from another branch. Removing.

arnaldo2792 commented on tools/testsys/src/run.rs:
```diff
@@ -266,55 +309,47 @@ pub(crate) struct Image {
         long = "ecs-resource-agent-image",
-        env = "TESTSYS_ECS_RESOURCE_AGENT_IMAGE",
-        default_value = "public.ecr.aws/bottlerocket-test-system/ecs-resource-agent:v0.0.2"
+        env = "TESTSYS_ECS_RESOURCE_AGENT_IMAGE"
```
It seems like default values were deleted.

ecpullen replied:
That is correct and expected. The images can be overwritten in the config file.

arnaldo2792 commented on tools/testsys/src/aws_resources.rs:
```diff
@@ -173,8 +206,12 @@ impl AwsK8s {
                     name: "eks-provider".to_string(),
-                    image: testsys_images.eks_resource.clone(),
-                    pull_secret: testsys_images.secret.clone(),
+                    image: testsys_images
+                        .eks_resource_agent_image
+                        .as_ref()
+                        .cloned()
+                        .unwrap(),
```
I think you could add a comment here explaining why this `unwrap` is safe. It took me a while to understand that the resulting `testsys_images` object will have a default value for `eks_resource_agent_image` regardless of the configurations provided by `Test.toml`.

Please do the same for any `unwrap` you left around that is safe to keep.

arnaldo2792 commented on tools/testsys/src/run.rs:
```diff
@@ -266,55 +309,47 @@ pub(crate) struct Image {
 
+/// This is a CLI parsable version of `testsys_config::TestsysImages`
 #[derive(Debug, Parser)]
 pub(crate) struct TestsysImages {
```
`CliConfig` has a 1 to 1 mapping with a different struct name (`GenerictVariantConfig`), would it make sense to do the same for this struct and either rename this or the counterpart in `testsys_config`?

ecpullen replied:
I don't think it quite works out, because the images available from the config can be split into 2 separate sections, controller and agents. I may add the controller image to the config file in a later pr.

bcressey commented on tools/testsys/Test.toml.example:
```diff
@@ -0,0 +1,78 @@
+# Configurable fields:
+# cluster-names: 
+#     For `aws-k8s` variants testing will be completed over all clusters listed.
+#     For `aws-ecs` variants, only the first cluster name will be used.
```
Why the different behavior? I don't think it's useful today but I suppose there could be cluster-level differences that we'd want to test in the future, and it seems harmless to allow it

bcressey commented on tools/testsys/Test.toml.example:
```diff
@@ -0,0 +1,78 @@
+# Configuration for only the `aws-k8s-1.21` variant (variant level configuration).
+["aws-k8s-1.21"]
+x86_64 = {}
+aarch64 = {kube-conformance-image = "<kube-conformance-image-uri>"}
```
nit: is the empty table needed?

I would also keep the same style for all the override tables, e.g. 
```suggestion
["aws-k8s-1.21".aarch64]
kube-conformance-image = "<kube-conformance-image-uri>"
```

bcressey commented on tools/testsys/Test.toml.example:
```diff
@@ -0,0 +1,78 @@
+aarch64 = {kube-conformance-image = "<kube-conformance-image-uri>"}
+
+# Configurable fields:
+# cluster-names: 
```
Can you include an example above of setting cluster names - maybe for `aws-k8s` with and without `-ipv6`?

bcressey commented on tools/testsys/Test.toml.example:
```diff
@@ -0,0 +1,78 @@
+# '{family}' configuration in `Test.toml`
+# '{platform}-{flavor}' configuration in `Test.toml`
+# '{platform}' configuration in `Test.toml`
+# TestSys default values
```
It might be easier to read the example if you provide each of them in priority order (from `variant` to `platform`) with a comment above each explaining which is which.

I'd also recommend picking a convention and sticking to it throughout, and specifically avoiding `{}` because of the overlap with handlebars.

Maybe `<USER-SPECIFIED-SHOUTING-KEBAB-CASE>` - there are some inconsistencies with `-` vs `_` (e.g. `aws-testing-role` but `account_id`).

bcressey commented on tools/testsys/Test.toml.example:
```diff
@@ -0,0 +1,78 @@
+[aws]
+agent-role = "<aws-testing-role>"
+
+# Configuration for all nvidia AWS variants on x86_64 (platform-flavor level configuration)
+[aws-nvidia.x86_64]
+instance-type = "p3.2xlarge"
```
You had mentioned maybe allowing a form like this:
```
[aws-nvidia]
instance-type = "p3.2xlarge"
```

Did you change your mind?

ecpullen replied:
We do allow that form; however we typically use different instance types based on the architecture.

bcressey commented on tools/testsys/Test.toml.example:
```diff
@@ -0,0 +1,78 @@
+agent-role = "arn:aws:iam::<account_id>:role/<role_name>"
+
+# To test sonobuoy certified conformance in other AWS partitions, sonobuoy e2e
+# images need to be available in a registry in the targeted region (cn-north-1, us-gov-west-1).
```
nit: this is not technically true if pulling images from ECR Public in GovCloud, and if reliability is not a concern then perhaps not for China either.

It might be better to document it as a way to set a custom conformance image, without reference to specific AWS partitions.

bcressey commented on tools/testsys/Test.toml.example:
```diff
@@ -0,0 +1,78 @@
+#     A map containing all secrets needed for resource creation and testing.
+# agent-role:
+#     The role that should be assumed by each test and resource agent.
+# kube-conformance-image: (K8s only)
+#     Specify a custom Kubernetes conformance image for E2E sonobuoy testing.
+# e2e-repo-registry: (K8s only)
+#     Configure the registry used for Sonobuoy E2E testing images.
```
Should we make these fields more general, e.g. `conformance-image` and `conformance-registry`?

ecpullen commented:
^ Address comments
* Rework `config_keys` vec creation to be more idiomatic
* Added comment to `config_keys` function
* Removed commented `MigrationsConfig`
* `merge` assign some variables before struct creation
* 'uri' -> 'URI'
* Removed unnecessary x86_64 table from 'Test.toml.example'
* 'selected' -> 'precedence'
* Provided configuration precedence for `aws-k8s-1.23-nvidia` variant on `x86_64`
* Provided a sample `cluster-names` for ipv6 and ipv4 testing with a disclaimer that ipv6 cluster creation is not yet supported.
* Changed `e2e-repo-registry` to `conformance-registry` to make it more general.
* Changed `kube-conformance-image` to `conformance-image` to make it more general.
* Reword cluster-names templated string comment.
* Comments for `unwrap()`'s
* Convert `cluster_name` to be `&str` to reduce cloning
* Added `debug` log if no cluster templates were provided
* Fixed inconsistencies with in user values


Still need to support multiple clusters with ecs.

ecpullen commented:
^ Enables ecs testing on multiple clusters and fixes an error in `TestsysImages::new()`
Tested with `cargo make -e BUILDSYS_VARIANT=aws-ecs-1 test` with the following configuration.
```yaml
[test]
repo="default"

[aws-ecs]
cluster-names = ["{{arch}}-{{variant}}", "{{arch}}-{{variant}}-dup"]
```

ecpullen commented:
^ Resolve merge conflict.

ecpullen commented:
^ Add small example config snippet to `TESTING.md` for instance type configuration.

ecpullen commented:
^ Update `Test.toml.example`'s `aws-k8s-1.23-nvidia` hierarchy example to also show what configuration level is being represented.

ecpullen commented:
^ Fixed 2 outdated comments related to testing on multiple ECS clusters.

stmcginnis reviewed (approved):
Looks fine to me, but would be good to get feedback from Ben and Arnaldo since they had previous comments.

Can you resolve the conversations that were addressed so it's clear what is done and what might still be an open question?

bcressey commented on tools/testsys/Test.toml.example:
```diff
@@ -0,0 +1,97 @@
+conformance-image = "<KUBERNETES-CONFORMANCE-IMAGE-URI>"
+
+# Configurable fields:
+# cluster-names: 
```
Should `instance-type` also be listed here? It might also make sense to put this before the "Configuration for all AWS variants" section, so that people see the possible values before the sample configurations.

bcressey commented on tools/testsys/src/run.rs:
```diff
@@ -101,16 +83,65 @@ pub(crate) struct Run {
+        let test_config = TestConfig::from_path_or_lock(&self.test_config_path, true)
+            .context("Unable to read test config")?;
+
+        let test = test_config.test.as_ref().cloned().unwrap_or_default();
```
nit: I'd prefer a more specific variable name, like
```suggestion
        let test_opts = test_config.test.as_ref().cloned().unwrap_or_default();
```

bcressey commented on tools/testsys/src/run.rs:
```diff
@@ -150,21 +186,31 @@ impl Run {
                 }
             });
 
+        let images = vec![
+            Some(self.agent_images.into()),
+            Some(test.testsys_images),
+            test.testsys_image_registry
+                .map(testsys_config::TestsysImages::new),
+            Some(testsys_config::TestsysImages::public_images()),
+        ]
+        .into_iter()
+        .flatten()
+        .fold(Default::default(), testsys_config::TestsysImages::merge);
```
The difference between `testsys_config` (a module) and `test_config` (a structure) is pretty key to understanding this, but I had them conflated in my mind for a while.

Maybe something like this to get rid of the ambiguity:
```
let (cfg_test_opts, cfg_variant_overrides) = TestConfig::from_path_or_lock(&self.test_config_path, true)
            .context("Unable to read test config")?.split();
```

Where `split` takes `Self` and destructures it into a tuple, so we don't need to keep a `TestConfig` around.

ecpullen replied:
I like that `reduced_config` is a function of `TestConfig` splitting would remove that feature since the other configs are just a `HashMap`.

bcressey commented on tools/testsys/src/aws_resources.rs:
```diff
@@ -173,23 +208,29 @@ impl AwsK8s {
+                        .cloned()
+                        // This is a safe unwrap because `testsys_images` has a default value set
+                        // for `eks_resource_agent`.
+                        .unwrap(),
```
nit: this is the same info but more useful to a developer that sees the theoretical panic:
```suggestion
                        .expect("missing default image for eks resource agent"),
```

bcressey commented on tools/testsys/src/aws_resources.rs:
```diff
@@ -390,25 +452,54 @@ impl AwsEcs {
         test: TestType,
         testsys_images: &TestsysImages,
```
Not for this PR - it would help to split out `AwsEcs` and `AwsK8s` into separate modules. This one is getting pretty long.

bcressey commented on tools/testsys-config/src/lib.rs:
```diff
@@ -0,0 +1,416 @@
+        toml::from_str(&test_config_str).context(error::InvalidTomlSnafu { path })
+    }
+
+    /// Deserializes a TestConfig from a Test.lock file at a given path
+    pub fn from_lock_path<P>(path: P) -> Result<Self>
+    where
+        P: AsRef<Path>,
+    {
+        let path = path.as_ref();
+        let test_config_str = fs::read_to_string(path).context(error::FileSnafu { path })?;
+        serde_yaml::from_str(&test_config_str).context(error::InvalidLockSnafu { path })
+    }
```
Do we need to support `Test.lock`?

IIRC for `Infra.toml` the idea was that `Infra.lock` would hold the concrete resource IDs as opposed to abstract values. If we don't have a similar need here, I'd just leave it out for now.

ecpullen replied:
Ok. I just tried to replicate `Infra.toml`. I can take it out.

bcressey commented on tools/testsys-config/src/lib.rs:
```diff
@@ -0,0 +1,416 @@
+    pub assume_role: Option<String>,
+}
+
+/// Create a vec if relevant keys for this variant ordered from most specific to least specific.
```
Nice, this is very lightweight for the power it offers.

```suggestion
/// Create a vec of relevant keys for this variant ordered from most specific to least specific.
```

bcressey commented on tools/testsys-config/src/lib.rs:
```diff
@@ -0,0 +1,416 @@
+impl GenericVariantConfig {
+    /// Overwrite the unset values of `self` with the set values of `other`
+    fn merge(self, other: Self) -> Self {
+        let cluster_names = if self.cluster_names.is_empty() {
+            other.cluster_names
+        } else {
+            self.cluster_names
+        };
+
+        let secrets = if self.secrets.is_empty() {
+            other.secrets
+        } else {
+            self.secrets
+        };
```
It is sort of tempting to have `cluster_names` and `secrets` be `Option`, and to have empty lists deserialize into `None`, just so all of the merge logic is the same. But I'm not sure it's worth it 😀

bcressey commented on tools/testsys-config/src/lib.rs:
```diff
@@ -0,0 +1,416 @@
+
+impl TestsysImages {
+    /// Create an images config for a specific registry.
+    pub fn new<S1>(registry: S1) -> Self
```
nit:
```suggestion
    pub fn new<S>(registry: S) -> Self
```

ecpullen commented:
^
* Adds `instance-types` to example config file
* Fixed typos in lib.rs
* Remove lock support for `Test.toml`
* Replace `unwrap`s with `expect`

ecpullen commented:
^ Change generic `S1` to `S`.

ecpullen commented:
^ Fix `cargo fmt`

ecpullen commented:
^ rebase to develop to fix merge conflict

etungsten commented on tools/testsys-config/src/lib.rs:
```diff
@@ -0,0 +1,364 @@
+    pub secrets: BTreeMap<String, SecretName>,
+    /// The role that should be assumed for this particular variant
+    pub assume_role: Option<String>,
+    /// The sonobuoy conformance image that should be used for this variant
+    pub kube_conformance_image: Option<String>,
```
nit: to clarify a bit more
```suggestion
    /// The kubernetes conformance image that should be used for this variant
    pub kube_conformance_image: Option<String>,
```

etungsten commented on tools/testsys/src/aws_resources.rs:
```diff
@@ -43,30 +43,57 @@ impl AwsK8s {
+                self.kube_variant(),
+            )?;
+            crds.append(&mut match &test {
+                TestType::Conformance => self.sonobuoy_test_crds(
+                    testsys_images,
+                    SonobuoyMode::CertifiedConformance,
+                    cluster_name,
+                )?,
+                TestType::Quick => {
+                    self.sonobuoy_test_crds(testsys_images, SonobuoyMode::Quick, cluster_name)?
+                }
```
Seeing this makes me wonder how expanding this list would look like. For example, if I want to run a different sonobuoy test mode like `conformance-lite`, do I add to the `TestType` enum? But `TestType` enum needs to be generic across all tests including ECS testing where `conformance-lite` doesn't make sense.

ecpullen replied:
We will add the `conformance-lite` test type and then report that it is unavailable for the other variants if a user tries to use them.

arnaldo2792 replied:
Isn't the `conformance-lite` test type in sonobuoy just a NGINX pod? If so, we could have the same `conformance-lite` testing for ECS and support both orchestrators.

etungsten replied:
You're thinking of the quick test mode. `conformance-lite` is a subset of tests from the certified K8s conformance tests suite that is faster to run.

etungsten commented on tools/testsys/src/aws_resources.rs:
```diff
@@ -730,13 +832,17 @@ trait Migration {
-                    pull_secret: testsys_images.secret.clone(),
+                    image: testsys_images
+                        .migration_test_agent_image
+                        .as_ref()
+                        .cloned()
```
This can probably be simplified to just a `cloned()` or `to_owned()`

etungsten commented on tools/testsys/src/aws_resources.rs:
```diff
@@ -497,21 +591,25 @@ impl AwsEcs {
-                    pull_secret: testsys_images.secret.clone(),
+                    image: testsys_images
+                        .ecs_resource_agent_image
+                        .as_ref()
+                        .cloned()
```
Is the `as_ref()` is necessary?
```suggestion
                        .cloned()
```

etungsten commented on tools/testsys/src/aws_resources.rs:
```diff
@@ -549,16 +651,20 @@ impl AwsEcs {
-                    pull_secret: testsys_images.secret.clone(),
+                    image: testsys_images
+                        .ec2_resource_agent_image
+                        .as_ref()
+                        .cloned()
```
Same here maybe?
```suggestion
                        .cloned()
```

etungsten commented on tools/testsys/src/run.rs:
```diff
@@ -101,16 +83,65 @@ pub(crate) struct Run {
+    #[clap(long, env = "TESTSYS_ASSUME_ROLE")]
+    assume_role: Option<String>,
+
+    /// Specify the instance type that should be used. This is only applicable for aws-* variants.
+    /// It can be omitted for non-aws variants and can be omitted to use default instance types.
+    #[clap(long, env = "TESTSYS_INSTANCE_TYPE")]
+    instance_type: Option<String>,
```
This needs to be updated once https://github.com/bottlerocket-os/bottlerocket-test-system/pull/602 merges I think

ecpullen replied:
It does not. We pin to a released version of testsys.

etungsten reviewed (commented):
[testsys.kubeconfig.eksctl.lock](https://github.com/bottlerocket-os/bottlerocket/pull/2393/files#diff-ea6827e4ef448d2a22b7c301c146c7419eb1a07c55a187c9f03cff1ff372fcd9) shouldn't be part of the PR.

Can you also give more details on how the testing was conducted, e.g how `test.toml` was set up and what the outputs look like?

webern commented:
Does `testsys.kubeconfig.eksctl.lock` need to be deleted and .gitignored?

webern commented on tools/testsys/Test.toml.example:
```diff
@@ -0,0 +1,99 @@
+eks-resource-agent-image = "public.ecr.aws/bottlerocket-test-system/eks_resource_agent:v0.0.2"
+
+# The test configuration is broken up into different levels depending on the variant.
+# Configurations will have the following precedence:
+# Arguments from the command line
+# Environment variables from the command line
```
```suggestion
# Arguments passed at the command line will take precedence over environment variables.
# Configurations will have the following resolution precedence:
```

webern commented on tools/testsys/Test.toml.example:
```diff
@@ -0,0 +1,99 @@
+# '<VARIANT>' configuration in `Test.toml`
+# '<FAMILY>-<FLAVOR>' configuration in `Test.toml`
+# '<FAMILY>' configuration in `Test.toml`
+# '<PLATFORM>-<FLAVOR>' configuration in `Test.toml`
```
Possible dumb question - How do you distinguish between a variant and `<PLATFORM>-<FLAVOR>`. For example, if my platform is `p` and my flavor is `f`, then `p-f` could be a platform-flavor or it could be a variant name. Note, I haven't read the code yet. I'm just thinking about a case where I have a platform `foo` and a flavor `bar`, e.g. in the variant `foo-x-1.1-bar`, but I also have a variant named `foo-bar`.

ecpullen replied:
If that happens then it will treat the `foo-bar` table as being variant level configuration. That is an edge case that I don't think we need to document.

webern commented on tools/testsys/Test.toml.example:
```diff
@@ -0,0 +1,99 @@
+# '<PLATFORM>' configuration in `Test.toml`
+# TestSys default values
+
+# For the `aws-k8s-1.23-nvidia` variant on `x86_64`, configurations will have the following precedence:
```
In the previous section you described variant, family, platform etc, without mentioning ARCH. Now we have ARCH without having introduced the concept. I'm left unsure how ARCH figures into the precedence scheme.

webern commented on tools/testsys/Test.toml.example:
```diff
@@ -0,0 +1,99 @@
+# TestSys default values
+
+# For the `aws-k8s-1.23-nvidia` variant on `x86_64`, configurations will have the following precedence:
+# Arguments from the command line
```
I don't quite understand why it says this here. In the previous section you talked about arguments from the command line taking precedence over environment variables. In this section we have the sentence fragment:

```text
# Arguments from the command line
```

webern commented on tools/testsys/Test.toml.example:
```diff
@@ -0,0 +1,99 @@
+cluster-names = ["{{arch}}-{{variant}}", "{{arch}}-{{variant}}-ipv6"]
+
+# A custom conformance registry may be needed for testing if image pull reliability is a concern.
+conformance-registry = "<ACCOUNT>.dkr.ecr.cn-north-1.amazonaws.com.cn"
```
Does every key in these "configuration" sections correspond to a testsys command line argument? If so, could that be more explicitly described in the overview at the beginning?

ecpullen replied:
That is a rule with come cli's but that is not guaranteed here, for instance `testsys` can only use a single cluster, but multiple clusters can be described in `Test.toml`.

webern commented on tools/testsys/Test.toml.example:
```diff
@@ -0,0 +1,99 @@
+["aws-k8s-1.21".aarch64]
+conformance-image = "<KUBERNETES-CONFORMANCE-IMAGE-URI>"
+
+# Configurable fields:
```
What is a "field"?

webern reviewed (changes requested):
I think if you land go to Test.toml.example to try and figure out what Test.toml is for and how it is used, you might be left confused. I've left a few comments that hopefully show what I mean.

webern commented on tools/testsys-config/src/lib.rs:
```diff
@@ -0,0 +1,364 @@
+
+    /// Deserializes a TestConfig from a given path, if it exists, otherwise builds a default
+    /// config
+    pub fn from_path_or_default<P>(path: P) -> Result<Self>
```
It might be helpful for a user to know (via stdout) whether they are getting their file loaded or getting a default. For example if I have a typo in the path that I pass `/home/me/bottlerocket-teesting/Test.toml`, I could be left scratching my head as to why my configuration file is not working.

ecpullen replied:
This issue the same loading behavior as `Infra.toml` if we are going to change one we should probably change both.

webern replied:
Hmmm, it should probably warn me there as well. I thought I remembered seeing a warning about a default Infra.toml. That might be because we generate one earlier in the build process when one does not exist.

arnaldo2792 replied:
I think there is a warning, as well as another warning when the default "development" keys are used to sign the repos.

webern commented on tools/testsys-config/src/lib.rs:
```diff
@@ -0,0 +1,364 @@
+    /// Create a single config for the `variant` and `arch` from this config. `starting_config` is
+    /// used as the config with the highest precedence. If no starting config is available,
+    /// `Default::default()` can be used.
+    pub fn reduced_config<S>(
```
`resolve_config`?

Can you better explain the purpose of the `starting_config` argument?

"Create a single config [...] from this config" is unclear.

Maybe it should be an `Option<GenericVariantConfig>`. It's not clear how to call the function if "no starting_config is available", but maybe what you're saying is that it is an optional argument.

ecpullen replied:
The technical term for the process is `reduce` so I'm not sure we want to change that. I have made `starting_config` an option.

webern commented on tools/testsys-config/src/lib.rs:
```diff
@@ -0,0 +1,364 @@
+#[derive(Debug, Default, Deserialize, Serialize, PartialEq, Eq, Clone)]
+#[serde(deny_unknown_fields)]
+#[serde(rename_all = "kebab-case")]
+pub struct TestsysImages {
```
We're hard-coding a lot of things for our "supported" variants. Perhaps this is fine and to be dealt with later when we work on out-of-tree builds.

ecpullen replied:
We decided that the only acceptable fields in a config are fields that we know about and implement.

webern commented on tools/testsys-config/src/lib.rs:
```diff
@@ -0,0 +1,364 @@
+
+#[derive(Debug, Default, Deserialize, Serialize, PartialEq, Eq, Clone)]
+#[serde(deny_unknown_fields)]
+pub struct GenericConfig {
```
Explain

webern commented on tools/testsys-config/src/lib.rs:
```diff
@@ -0,0 +1,364 @@
+#[derive(Debug, Default, Deserialize, Serialize, PartialEq, Eq, Clone)]
+#[serde(deny_unknown_fields)]
+#[serde(rename_all = "kebab-case")]
+pub struct GenericVariantConfig {
```
Explain

webern commented on tools/testsys-config/src/lib.rs:
```diff
@@ -0,0 +1,364 @@
+    pub sonobuoy_test_agent_image: Option<String>,
+    pub ecs_test_agent_image: Option<String>,
+    pub migration_test_agent_image: Option<String>,
+    pub testsys_agent_pull_secret: Option<String>,
```
Can I have one agent container that is published in a different place? Or would I need to publish all of these containers to the same container registry?

ecpullen replied:
They can be in different ones. There is a note about it in `Test.toml.example`.

webern commented on tools/testsys-config/src/lib.rs:
```diff
@@ -0,0 +1,364 @@
+        S: Into<String>,
+    {
+        let arch = arch.into();
+        config_keys(variant)
+            .into_iter()
+            .filter_map(|key| self.configs.get(&key).cloned())
+            .flat_map(|config| vec![config.for_arch(&arch), config.config])
+            .fold(starting_config, GenericVariantConfig::merge)
```
Since is the core of the matching algorithm, perhaps a description of how it works could go here in case one's (ahem) functional programming in Rust skills are lacking.

webern commented on tools/testsys/src/aws_resources.rs:
```diff
@@ -285,14 +336,31 @@ impl AwsK8s {
+                                .map(|e2e_registry| {
+                                    base64::encode(format!(
+                                        r#"buildImageRegistry: {e2e_registry}
+dockerGluster: {e2e_registry}
```
Is `dockerGluster` correct?

ecpullen replied:
Yes.

webern commented on tools/testsys/src/aws_resources.rs:
```diff
@@ -10,24 +10,24 @@ use aws_sdk_ec2::Region;
+    rendered_cluster_name, AwsEcsVariantConfig, AwsK8sVariantConfig, TestsysImages,
+};
 
 pub(crate) struct AwsK8s {
```
I'll leave this file's large diff alone, but I'm flagging it as another place that we will have to think about extensibility for out-of-tree variants.

ecpullen commented:
^ 
* Added `testsys.kubeconfig.eksctl.lock` to `.gitignore`
* Change comment for `kube_conformance_image`
* Convert `.as_ref().cloned()` to `to_owned()`
*  Reworded precedence comment in `Test.toml.example`
* Added a note describing architecture configuration in `Test.toml.example`
* Removed `arguments from command line` from precedence lists and added a not about them instead.
*  Changed `Configurable fields` to `Configurable values`
* Added doc comment for `reduced_config` function
* Added comments documenting the transformations occurring in `reduced_config`
* Added doc comment for `GenericConfig` and `GenericVariantConfig`

webern commented on tools/testsys/Test.toml.example:
```diff
@@ -0,0 +1,103 @@
+# useful to create a repo in `Infra.toml` that contains the infrastructure
+# needed for testing
+repo = "default"
+# The registry containing alternate TestSys agent images
```
Is "alternate" correct in this sentence?

```suggestion
# The registry containing TestSys agent images
```

webern commented on tools/testsys/Test.toml.example:
```diff
@@ -0,0 +1,103 @@
+# Arguments passed at the command line will take precedence over environment variables and any
+# values specified in the configuration file.
+# Configurations will have the following resolution precedence:
+# '<VARIANT>' configuration in `Test.toml`
+# '<FAMILY>-<FLAVOR>' configuration in `Test.toml`
+# '<FAMILY>' configuration in `Test.toml`
+# '<PLATFORM>-<FLAVOR>' configuration in `Test.toml`
+# '<PLATFORM>' configuration in `Test.toml`
```
I'm not sure if it should be demonstrated in this fashion, but do I understand this correctly?

(Also, is " configuration in `Test.toml`" actually required after each of these? If not, i.e. if it's obvious, then it would be less cluttered without that repeated suffix)

```suggestion
# '<VARIANT>'.ARCH
# '<VARIANT>'
# '<FAMILY>-<FLAVOR>'.ARCH
# '<FAMILY>-<FLAVOR>'
# '<FAMILY>'.ARCH
# '<FAMILY>'
# '<PLATFORM>-<FLAVOR>'.ARCH
# '<PLATFORM>-<FLAVOR>'
# '<PLATFORM>'.ARCH
# '<PLATFORM>'
```

etungsten commented on .gitignore:
```diff
@@ -6,7 +6,9 @@
 /Infra.toml
+/Test.toml
 /testsys.kubeconfig
+/testsys.kubeconfig.eksctl.lock
```
I don't think we should add this here. How is this lock file generated? From my experience, it's only when I create a new cluster with `eksctl`. This lock file only shows up under pretty specific use-cases (i.e. using eksctl to create the testsys cluster in the root dir of this repo).

edit: I believe this is a bug in eksctl that eventually will be fixed: https://github.com/weaveworks/eksctl/issues/5698/. In any case, I don't think an eksctl-specific lock file should be added here

webern replied:
>  I don't think an eksctl-specific lock file should be added here

I've never really understood the aversion to putting things in .gitignore.

arnaldo2792 replied:
For the lack of a better place to put this, `testsys.kubeconfig.eksctl.lock` was added in the first commit but deleted in the second commit. Please clean up the file so that it doesn't show up in either commit.

etungsten replied:
Please remove `/testsys.kubeconfig.eksctl.lock` from `.gitignore`. It's not specific to Bottlerocket or any of its build/test frameworks. I see this akin to adding other developer workflow specific things like IDE configuration to the project's `.gitignore`.

For developers who don't want `eksctl.lock` files to show up in any of their git trees, they can use their personal global/local `.gitignore`. See https://git-scm.com/docs/git-config#Documentation/git-config.txt-coreexcludesFile

webern commented on TESTING.md:
```diff
@@ -118,6 +118,24 @@ The Bottlerocket components are found in the `testsys-bottlerocket-aws` Kubernet
 
+### Configuration
+
+There are many arguments that can be configured via environment variables with `cargo make`; however, it is possible to create a configuration file instead.
+Check out the [example config file](tools/testsys/Test.toml.example) for a sample `Test.toml` file.
```
I stumbled on this a little bit because I was thinking about a configuration file for `cargo make` (like an `.env` file). I think it could be said a bit more directly. Here's an attempt:

```suggestion
Tests can be configured either by passing environment variables to `cargo make`, or by creating a `Test.toml` configuration file.
Check out the [example config file](tools/testsys/Test.toml.example) for a sample `Test.toml` file.
```

webern commented on TESTING.md:
```diff
@@ -118,6 +118,24 @@ The Bottlerocket components are found in the `testsys-bottlerocket-aws` Kubernet
+There are many arguments that can be configured via environment variables with `cargo make`; however, it is possible to create a configuration file instead.
+Check out the [example config file](tools/testsys/Test.toml.example) for a sample `Test.toml` file.
+
+To support different instance types on regular `aws-k8s` variants and `nvidia` `aws-k8s` variants, the following configuration could be used.
```
Sorry to be so nit picky with the documentation... something about the word "regular" doesn't sit right. Would it work to let the example speak for itself?

(Also, this PR confirms my instinct that we want a newline to precede code blocks https://github.com/bottlerocket-os/bottlerocket/pull/2283/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L490)

```suggestion
For example, we can specify different instance types based on variant requirements:

```

arnaldo2792 replied:
Nit on the nit: use "you" instead of "we".

webern commented on tools/testsys-config/src/lib.rs:
```diff
@@ -0,0 +1,382 @@
+    pub testsys_agent_pull_secret: Option<String>,
+}
+
+const AGENT_VERSION: &str = "v0.0.2";
```
We shouldn't block on this, but I'm noticing that the version string `0.0.2` exists in a few different places. I wonder if we can consolidate it later.

Edit: I think this is related https://github.com/bottlerocket-os/bottlerocket-test-system/issues/542

webern commented on tools/testsys/Test.toml.example:
```diff
@@ -0,0 +1,103 @@
+# This is an example testing configuration for TestSys, the tool that
```
I tried to see if I could make any improvements, and it was easier to do in an editor than in the PR. Some things I tried:
- The wrapping seemed off, so I set a ruler at 100 and re-wrapped it.
- I re-wrote some things because it felt like some context was missing. 
- I added some newlines and removed some repeated text to let it breathe
- It felt like architecture was missing from the explanation of precedence (as noted in another comment), so I tried incorporating it.

I don't want to block this important PR, but did want to take a stab improvement so spent some time on it.

```toml
# This is an example testing configuration for TestSys, the tool that is used to validate
# Bottlerocket builds.

# This section contains configuration details for all testing
[test]

# The repo from `Infra.toml` that should be used for Bottlerocket update images. It may be useful to
# create a repo in `Infra.toml` that contains the infrastructure needed for testing
repo = "default"

# The registry containing alternate TestSys agent images
testsys-images-registry = "public.ecr.aws/bottlerocket-test-system"

# The URI for the EKS resource agent that should be used. An individual agent's provided URI will be
# used even if `testsys-images-registry` is present.
eks-resource-agent-image = "public.ecr.aws/bottlerocket-test-system/eks_resource_agent:v0.0.2"

# Test Configurations
#
# Testing requirements tend to differ by variant and architecture. This configuration file provides
# the ability to set values that apply generally to a broad group of similar variants, and to
# override those values at a more granular level. For example, you can set a value for all `aws-k8s`
# variants, then override that value for 'aws-k8s-nvidia' variants, and further override the value
# for 'aws-k8s-nvidia'.aarch64 builds.
#
# The mechanism for resolving configuration values has the following order of precedence:
#
# '<VARIANT>'.ARCH
# '<VARIANT>'
# '<FAMILY>-<FLAVOR>'.ARCH
# '<FAMILY>-<FLAVOR>'
# '<FAMILY>'.ARCH
# '<FAMILY>'
# '<PLATFORM>-<FLAVOR>'.ARCH
# '<PLATFORM>-<FLAVOR>'
# '<PLATFORM>'.ARCH
# '<PLATFORM>'
#
# For concrete example, given a variant such as `aws-k8s-1.23-nvidia` with the architecture
# `x86_64`, configurations will have the following order of precedence:
# ['aws-k8s-1.23-nvidia'.x86_64]
# ['aws-k8s-1.23-nvidia']
# ['aws-k8s-nvidia'.x86_64]
# ['aws-k8s-nvidia']
# ['aws-k8s'.x86_64]
# ['aws-k8s']
# ['aws-nvidia'.x86_64]
# ['aws-nvidia']
# ['aws'.x86_64]
# ['aws']
#
# Note: values passed by command line argument will take precedence over those passed by environment
# variable, and both take precedence over values set by `Test.toml`.

# Example Configurations

# Configuration for all variants with the `aws` platform.
[aws]
agent-role = "<AWS-TESTING-ROLE>"

# Configuration for all nvidia AWS variants on x86_64 (platform-flavor level configuration)
[aws-nvidia.x86_64]
instance-type = "p3.2xlarge"

# Configuration for all nvidia AWS variants on aarch64 (platform-flavor level configuration)
[aws-nvidia.aarch64]
instance-type = "g5g.2xlarge"

# Configuration for all `aws-k8s` variants testing (family level configuration).
[aws-k8s]
# A single role can be assumed by agents to test all `aws-k8s` variants in a separate
# testing account.
agent-role = "arn:aws:iam::<ACCOUNT-ID>:role/<ROLE-NAME>"

# The cluster name templating can be defined for all `aws-k8s` variants. To test on ipv4 and ipv6
# clusters, the following templates could be used. Note: TestSys does not currently support creating
# ipv6 clusters, so the ipv6 cluster must already exist.
cluster-names = ["{{arch}}-{{variant}}", "{{arch}}-{{variant}}-ipv6"]

# A custom conformance registry may be needed for testing if image pull reliability is a concern.
conformance-registry = "<ACCOUNT>.dkr.ecr.cn-north-1.amazonaws.com.cn"

# If testing using a kind cluster, AWS credentials need to be passed as a K8s secret.
secrets = {"awsCreds" = "myAwsCredentials"}

# Configuration for all nvidia AWS variants on x86_64 (family-flavor level configuration)
[aws-ecs-nvidia.x86_64]
instance-type = "p3.2xlarge"

# Configuration for all nvidia AWS variants on aarch64 (family-flavor level configuration)
[aws-ecs-nvidia.aarch64]
instance-type = "g5g.2xlarge"

# Configuration for only the `aws-k8s-1.21` variant (variant level configuration).
["aws-k8s-1.21".aarch64]
conformance-image = "<KUBERNETES-CONFORMANCE-IMAGE-URI>"

# Configurable values:
#
# cluster-names: 
#     All clusters the variant should be tested over. Cluster naming supports templated strings, and
#     both `arch` and `variant` are provided as variables (`{{arch}}-{{variant}}`).
#
# instance-type:
#     The instance type that should be used for testing.
#
# secrets:
#     A map containing all secrets needed for resource creation and testing.
#
# agent-role:
#     The role that should be assumed by each test and resource agent.
#
# conformance-image: (K8s only)
#     Specify a custom image for conformance testing. For `aws-k8s` variants this will be used as a
#     custom Kubernetes conformance image for Sonobuoy.
#
# conformance-registry: (K8s only)
#     Specify a custom registry for conformance testing images.
#     For `aws-k8s` variants this will be used as the Sonobuoy e2e registry.
```

arnaldo2792 replied:
It is odd to me to document in the example configuration, rather than in the configuration file. I have not seen this convention in some of the open source projects I use, so probably we should consider stop documenting in example configuration files, and write down proper documentation where it should be. (Not for this PR)


Regarding the formatting in the toml file, would it make sense to use a linting tool and add that linting to the PR checks, and fix whatever we have to in other example configuration files? (Not in this PR)

webern replied:
I doubt if a TOML linter would introspect the formatting of comments, which is where I had issues. I agree that maybe some of this could/maybe-should go in documentation instead of the example config. However it might actually be easier to find in the configuration. I know documentation discoverability is an area we want to improve in the project.

webern reviewed (commented):
Ok, I've totally finished reviewing this. Great work!

arnaldo2792 commented on TESTING.md:
````diff
@@ -118,6 +118,24 @@ The Bottlerocket components are found in the `testsys-bottlerocket-aws` Kubernet
+instance-type = "g5g.2xlarge"
+```
+
+Since `aws-k8s-nvidia` which is a `<FAMILY>-<FLAVOR>` level configuration it will take precedence over `aws-k8s` which is `<FAMILY>` level configuration.
````
```suggestion
Since `aws-k8s-nvidia` is a `<FAMILY>-<FLAVOR>` level configuration it will take precedence over `aws-k8s` which is `<FAMILY>` level configuration.
```

arnaldo2792 commented on tools/testsys/src/aws_resources.rs:
```diff
@@ -81,15 +108,23 @@ impl AwsK8s {
-        let ec2 = self.ec2_crd(testsys_images, Some(ami))?;
+        let eks = self.eks_crd(cluster_name, testsys_images)?;
+        let ec2 = self.ec2_crd(cluster_name, testsys_images, Some(ami))?;
+        let instance_provider = ec2.name().unwrap();
```
In a previous forced push you replaced `unwrap` with `expect`, but you missed two, I left a comment in the other.

arnaldo2792 commented on tools/testsys/src/aws_resources.rs:
```diff
@@ -426,29 +502,41 @@ impl AwsEcs {
-        let ec2 = self.ec2_crd(testsys_images, Some(ami))?;
+        let ecs = self.ecs_crd(cluster_name, testsys_images)?;
+        let ec2 = self.ec2_crd(cluster_name, testsys_images, Some(ami))?;
+        let instance_provider = ec2.name().unwrap();
```
Here is the other `unwrap` to be replaced with `expect`.

ecpullen commented:
^ 
* Remove `eksctl.lock` from history
* Updated `TESTING.md` with comments
* Added warning for use of default `Test.toml`
* Used the proposed `Test.toml.example`
* Replaced`unwrap()` with `expect()`

webern reviewed (approved):
:star: :rocket: :tada: :star:
